README.md
# LogDNA LogSpout
A Docker LogSpout image to stream logs from your containers to LogDNA.

## How to Use

### Environment Variables
The following variables can be used to tune `LogSpout` for specific use cases.

#### Log Router Specific
* __FILTER_NAME__: Filter by Container Name with Wildcards, *Optional*
* __FILTER_ID__: Filter by Container ID with Wildcards, *Optional*
* __FILTER_SOURCES__: Filter by Comma-Separated List of Sources, *Optional*
* __FILTER_LABELS__: Filter by Comma-Separated List of Labels, *Optional*

__Note__: More information can be found [here](https://github.com/gliderlabs/logspout/tree/0da75a223db992cd5abc836796174588ddfc62b4/routesapi#routes-resource).

#### Ingestion Specific
* __LOGDNA_KEY__: LogDNA Ingestion Key, *Required*
* __HOSTNAME__: Alternative Hostname, *Optional*
  * __Default__: System's Hostname
* __LOGDNA_URL__: Specific Endpoint to Stream Log into, *Optional*
  * __Default__: `logs.logdna.com/logs/ingest`
* __TAGS__: Comma-Separated List of Tags, *Optional*

__Note__: Logging the `LogSpout` Container is recommended to keep track of HTTP Request Errors or Exceptions.

#### Limits
* __FLUSH_INTERVAL__: How frequently batches of logs are sent (in `milliseconds`), *Optional*
  * __Default__: 250
* __HTTP_CLIENT_TIMEOUT__: Time limit (in `seconds`) for requests made by this HTTP Client, *Optional*
  * __Default__: 30
  * __Source__: [net/http/client.go#Timeout](https://github.com/golang/go/blob/master/src/net/http/client.go#L89-L104)
* __INACTIVITY_TIMEOUT__: How long to wait for inactivity before declaring failure in the `Docker API` and restarting, *Optional*
  * __Default__: 1m
  * __Note__: More information about the possible values can be found [here](https://github.com/gliderlabs/logspout#detecting-timeouts-in-docker-log-streams). Also see [`time.ParseDuration`](https://golang.org/pkg/time/#ParseDuration) for valid format as recommended [here](https://github.com/gliderlabs/logspout/blob/e671009d9df10e8139f6a4bea8adc9c7878ff4e9/router/pump.go#L112-L116).
* __MAX_BUFFER_SIZE__: The maximum size (in `mb`) of batches to ship to `LogDNA`, *Optional*
  * __Default__: 2
* __MAX_REQUEST_RETRY__: The maximum number of retries for sending a line when there are network failures, *Optional*
  * __Default__: 5

### Docker
Create and run container named *logdna* from this image using CLI:
```bash
sudo docker run --name="logdna" --restart=always \
-d -v=/var/run/docker.sock:/var/run/docker.sock \
-e LOGDNA_KEY="<LogDNA Ingestion Key>" \
logdna/logspout:latest
```

### Docker Cloud
Append the following to your Docker Cloud stackfile:
```yaml
logdna:
  autoredeploy: true
  deployment_strategy: every_node
  environment:
    - LOGDNA_KEY="<LogDNA Ingestion Key>"
    - TAGS='{{.Container.Config.Hostname}}'
  image: 'logdna/logspout:latest'
  restart: always
  volumes:
    - '/var/run/docker.sock:/var/run/docker.sock'
```

### Elastic Container Service (ECS)
Modify your ECS Cloud Configuration file to have `LogDNA` Service as described below:
```yaml
services:
  logdna:
    environment:
        - LOGDNA_KEY="<LogDNA Ingestion Key>"
        - TAGS='{{ if .Container.Config.Labels }}{{index .Container.Config.Labels "com.amazonaws.ecs.task-definition-family"}}:{{index .Container.Config.Labels "com.amazonaws.ecs.container-name"}}{{ else }}{{.ContainerName}}{{ end }}'
    image: logdna/logspout:latest
    restart: always
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    deploy:
      mode: global
```

### Rancher
Modify your Rancher Compose Stackfile to have `LogDNA` Service as described below:
```yaml
version: '2'
services:
  logdna:
    image: logdna/logspout:latest
    environment:
      LOGDNA_KEY="<LogDNA Ingestion Key>"
    restart: always
    labels:
      io.rancher.container.hostname_override: container_name
      io.rancher.container.pull_image: always
      io.rancher.os.scope: system
    volumes:
    - /var/run/docker.sock:/tmp/docker.sock
```

### Docker Swarm
Modify your Docker Swarm Compose file to have `LogDNA` Service as described below:
```yaml
version: "3"
networks:
  logging:
services:
  logdna:
    image: logdna/logspout:latest
    networks:
      - logging
    volumes:
      - /etc/hostname:/etc/host_hostname:ro
      - /var/run/docker.sock:/var/run/docker.sock
    environment:
      - LOGDNA_KEY="<LogDNA Ingestion Key>"
    deploy:
      mode: global
```

### Notes
Do not forget to add `-u root` (in CLI) or `user: root` (in YAML) in case of having permission issues.

## Contributing
Contributions are always welcome. See the [contributing guide](/CONTRIBUTING.md) to learn how you can help. Build instructions for the agent are also in the guide.


CHANGELOG.md
# Change Log

This file documents all notable changes in `LogDNA LogSpout Image`. The release numbering uses [semantic versioning](http://semver.org).

## v1.3.2 - Released on July 25, 2023
* Changed auth mechinism to use `apiKey` instead of basic auth
    * this makes it compatible with pipeine ingest

## v1.3.1 - Released on April 21, 2021
* Removes a file that was erroneously committed to the repository

## v1.3.0 - Released on April 20, 2021
* Add a Makefile to do builds for multiple architectures (amd64, arm64)
    * Correct CI workflow to build from a `Makefile`
    * Use `docker manifest` to control tagging

## v1.2.0 - Released on February 20, 2020
* Update License
* Update Release Pipeline
* Remove Log Line Sanitization
* Update Making HTTP POST Requests and Retry Mechanism
* Add User-Agent Header to the POST Requests
* Update Authentication Mechanism

## v1.1.0 - Released on December 06, 2018
* Get Tags from Templates

## v1.0.0 - Released on August 29, 2018
* Initial Release


logdna/logdna.go
// Package logdna is LogSpout Adapter to Forward Logs to LogDNA
package logdna

// The MIT License (MIT)
// =====================
//
// Copyright (c) 2020 LogDNA, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import (
	"errors"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/gliderlabs/logspout/router"
	"github.com/gorilla/mux"
	"github.com/logdna/logspout/logdna/adapter"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

/*
   Common Functions
*/

// Getting Uint Variable from Environment:
func getUintOpt(name string, dfault uint64) uint64 {
	if result, err := strconv.ParseUint(os.Getenv(name), 10, 64); err == nil {
		return result
	}
	return dfault
}

// Getting Duration Variable from Environment:
func getDurationOpt(name string, dfault time.Duration) time.Duration {
	if result, err := strconv.ParseInt(os.Getenv(name), 10, 64); err == nil {
		return time.Duration(result)
	}
	return dfault
}

// Getting String Variable from Environment:
func getStringOpt(name, dfault string) string {
	if value := os.Getenv(name); value != "" {
		return value
	}
	return dfault
}

// metrics returns a http.Handler for the prometheus metrics
func metrics() http.Handler {
	r := mux.NewRouter()
	r.Handle("/metrics", promhttp.Handler())

	return r
}

func init() {
	router.HTTPHandlers.Register(metrics, "metrics")
	router.AdapterFactories.Register(NewLogDNAAdapter, "logdna")

	filterLabels := make([]string, 0)
	if filterLabelsValue := os.Getenv("FILTER_LABELS"); filterLabelsValue != "" {
		filterLabels = strings.Split(filterLabelsValue, ",")
	}

	filterSources := make([]string, 0)
	if filterSourcesValue := os.Getenv("FILTER_SOURCES"); filterSourcesValue != "" {
		filterSources = strings.Split(filterSourcesValue, ",")
	}

	r := &router.Route{
		Adapter:       "logdna",
		FilterName:    getStringOpt("FILTER_NAME", ""),
		FilterID:      getStringOpt("FILTER_ID", ""),
		FilterLabels:  filterLabels,
		FilterSources: filterSources,
	}

	if err := router.Routes.Add(r); err != nil {
		log.Fatal("Cannot Add New Route: ", err.Error())
	}
}

// NewLogDNAAdapter creates adapter:
func NewLogDNAAdapter(route *router.Route) (router.LogAdapter, error) {
	logdnaKey := os.Getenv("LOGDNA_KEY")
	if logdnaKey == "" {
		return nil, errors.New("Cannot Find Environment Variable \"LOGDNA_KEY\"")
	}

	if os.Getenv("INACTIVITY_TIMEOUT") == "" {
		os.Setenv("INACTIVITY_TIMEOUT", "1m")
	}

	config := adapter.Configuration{
		BackoffInterval:   getDurationOpt("HTTP_CLIENT_BACKOFF", 2) * time.Millisecond,
		FlushInterval:     getDurationOpt("FLUSH_INTERVAL", 250) * time.Millisecond,
		Hostname:          os.Getenv("HOSTNAME"),
		HTTPTimeout:       getDurationOpt("HTTP_CLIENT_TIMEOUT", 30) * time.Second,
		JitterInterval:    getDurationOpt("HTTP_CLIENT_JITTER", 5) * time.Millisecond,
		LogDNAKey:         logdnaKey,
		LogDNAURL:         getStringOpt("LOGDNA_URL", "logs.logdna.com/logs/ingest"),
		MaxBufferSize:     getUintOpt("MAX_BUFFER_SIZE", 2) * 1024 * 1024,
		RequestRetryCount: getUintOpt("MAX_REQUEST_RETRY", 5),
		Tags:              os.Getenv("TAGS"),
	}

	return adapter.New(config), nil
}
